docs/spa/index.md

```diff
@@ -0,0 +1,183 @@
+
+```js
+let fragment = window.location.hash.substring(1)
+let params = new URLSearchParams(fragment)
+let code = params.get("code")
+let error = params.get("error")
+
+window.location.replace("#")
+```
+
+The final line is not required, but can make for a nicer user experience.
+Replacing the location will remove the code from the browsers navigation bar.
+
+### Fetch User Data
+
+#### Option 1. Using backend endpoint
+
+**The CLIENT_SECRET for you application must not be used in your front end.**
+Send a request to your backend containg the code extracted in the previous section.
+Call DID.app's token endpoint to exchange your code for an id token.
+
+```js
+const fetch = require("node-fetch");
+
+let response = await fetch("https://auth.did.app/oidc/token", {
+  method: "POST",
+  body: `client_id=[CLIENT_ID]&client_secret=[CLIENT_SECRET]&code=${code}`
+});
+let data = response.json();
+```
+
+The data will contain your users information.
+
+<a name="pkce"></a>
+#### Option 2. Proof Key for Code Exchange (PKCE)
+
+Using PKCE allows your client application to retrieve the id token from a code directly, without exposing your client secret.
+To use PKCE you need to send a `code_challenge` with your authorization request.
+
+Generate a code challenge and code verifier.
+
+See [code samples](#generating-pkce) at the end for implementation of the functions used here.
+
+```js
+let codeVerifier = randomUrlBase64(32)
+let codeChallenge = await hashUrlBase64(codeVerifier)
+```
+
+The code challenge and code challenge method is sent with the authorization request.
+The code challenge method is `sha256`, it is the only value currently supported.
+
+Now your client can make a call to the token endpoint, instead of sending a client secret, it instead sends the code verifer.
+
+```js
+let response = await fetch("https://auth.did.app/oidc/token", {
+  method: "POST",
+  body: `client_id=[CLIENT_ID]&code_verifier=${code_verifier}&code=${code}`
+});
+let data = response.json();
+```
+
+The data will contain your users information.
+
+### Using user information
+
+In both cases the returned payload will contain an id_token, access_token and userinfo.
+
+```json
+{
+  "userinfo": {
+    "sub": "00000000-0000-0000-0000-000000000000",
+    "email": "user@example",
+    "email_verified": true
+  },
+  "id_token": "COMPACT JWT",
+  "access_token": "COMPACT JWT"
+}
+```
+
+The `sub` is the identifier of a user within your application.
+It is intended to be used as the primary identifier of your user and does not change.
+
+You should not rely on a users email being constant over time.
+
+At this point authentication is complete.
+It is up to your service to handle things from here.
+
+For backend applications this will involve starting a session for the authenticated user.
+How this is done will depend on the platform, or framework that you are using.
+
+<a name="generating-pkce"></a>
+### Generating code challenge and verifier
+
```

Comment:
What is this section trying to achieve?  I felt like I hadn't understood the relevance of this bit.

Reply:
This is basically an appendix to where the function `randomUrlBase64` is used earlier in the post

Reply:
Ah right ok. This wasn't clear is my feedback but now I know it's an appendix then that's quite helpful. Worth making it more explicit?

Reply:
How, would you do that?